### PR TITLE
Add Metadata type to AbortError & return validationErrors as metadata

### PR DESCRIPTION
**Pull request by simonedelmann**
### Metadata type
Adds encodable `Metadata` type, which can be used to encode any value. 

```swift
typealias Metadata = [String: MetadataValue]

struct MetadataValue: Encodable {
    var value: Any
}
```

**Please notice!** If any value will be stored which cannot be encoded, this will throw an error.

### Metadata property in AbortError
Adds a metadata property to `AbortError`, defaulting to `[:]`. This can be used to add any metadata for a special error.

```swift
extension FooError: AbortError  {
    var metadata: Metadata {
        return ["foo", .init("bar")]
    }
}
```

### Metadata for ValidationsErrors 
Returns individual validation errors via metadata property. The response for a `ValidationsError` looks like this:

```json
{
  "error": true,
  "metadata": {
    "validationErrors": {
      "email": "email is not a valid email address"
    }
  },
  "reason": "email is not a valid email address"
}
```

#### Special cases
Multiple validations for one key will be combined using ' and ' if they are declared at once.
```swift
validations.add("foo", as: String.self, is: .count(3...) && .alphanumeric)
```
will result in
```json
{
  "foo": "foo is less than minimum of 3 character(s) and should contain only alphanumeric characters"
}
```

Multiple validations declared individually will be returned as array.
```swift
validations.add("foo", as: String.self, is: .count(3...))
validations.add("foo", as: String.self, is: .alphanumeric)
```
will result in
```json
{
  "foo": [
    "foo is less than minimum of 3 character(s)",
    "foo should contain only alphanumeric characters"
  ]
}
```

Nested validations will be returned as object.
```swift
validations.add("foo") { foo in
    foo.add("bar", as: String.self, is: .count(3...))
}
```
will result in 
```json
{
  "foo": {
    "bar": "bar is less than minimum of 3 character(s)"
  }
}
```

### CODE REVIEW

**Comment by simonedelmann** [comment]
This PR is not ready to merge as the returned validation errors look like this:
`{"validationErrors":"[\"email\": [\"email is not a valid email address\"]]"}`
They should look like this, without escaped quotes:
`{"validationErrors":"["email": ["email is not a valid email address"]]"}`

File: Sources/Vapor/Utilities/Metadata.swift [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.singleValueContainer()
+
+        switch self {
+        case .string(let value):
+            try container.encode(value)
+        case .stringConvertible(let value):
+            try container.encode(value.description)
+        case .dictionary(let value):
+            try container.encode(value)
```
Comment (tanner0101) [inline Sources/Vapor/Utilities/Metadata.swift]: This should be:
```swift
let keyedContainer = container.nestedContainer(keyedBy: SomeCodingKey.self)
value.forEach { (key, value) in 
    keyedContainer.encode(value, forKey: key)
}
```

File: Sources/Vapor/Utilities/Metadata.swift [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        case .array(let value):
+            try container.encode(value)
```
Comment (tanner0101) [inline Sources/Vapor/Utilities/Metadata.swift]: Same here except using unkeyed containers.

Comment (simonedelmann) [reply Sources/Vapor/Utilities/Metadata.swift]: Updated!

File: Sources/Vapor/Utilities/Metadata.swift (hunk 1 above)
Comment (simonedelmann) [reply Sources/Vapor/Utilities/Metadata.swift]: Updated, although I had to change some bits.

**Comment by simonedelmann** [comment]
@tanner0101 Thanks for your help, I've updated the encoder! Unfortunately still the quotes get escaped. Maybe you can have a look at that?

**Comment by simonedelmann** [comment]
Finally it works, not using an `enum` but a `struct`. This whole thing should be non-breaking.

**Comment by simonedelmann** [comment]
I have included support for nested validations.

**Comment by Joannis** [comment]
@simonedelmann is the PR ready now? Earlier you stated that it's still WIP.

**Comment by simonedelmann** [comment]
@Joannis Short answer: Yes, it is ready to merge!

Long answer:
I made two PRs: #2241 and #2238.
- #2241 is adding validation errors as metadata
- #2238 is adding a new type of validations (validations for nested collections)

If **both** will get merged, it will be needed to add metadata for nested collection validations within a new PR.
If **only one of them** will get merged, I can add the missing connecting part to the other PR.

**Comment by simonedelmann** [comment]
Thanks for your attention! Unfortunately I won't have time to revisit the PR in the next few weeks, so feel free to do whatever you want with the PR.

**Comment by Lloople** [comment]
> Thanks for your patience @simonedelmann!
> 
> I'm hesitant about adding more complexity to Vapor's errors but I think this could be quite helpful for users. I think a good compromise here could by making this opt-in with something like:
> 
> ```swift
> app.validation.enableErrorMetadata = true
> ```

From my experience, I think it would be better to always send this information when a validation failed. Forms are not usually really long and the developer can choose if they want to show these errors to the user or not

**Pull request closed without merging**